### PR TITLE
New ESLint rules

### PR DESCRIPTION
**Pull request by pfulop**
### Summary:
This PR adds more lint rules so uniform code style is more easily achievable. It also ads pre-hook to disallow commit if there is a lint error on staged files and also hook to run tests before pushing.

I also run initial lint on the codebase.

### GitHub linked issue:
#896 

### Type (select 1):
- [ ] Bug fix
- [X] Feature implementation
- [ ] Documentation improvement
- [ ] Testing improvement

- [ ] Test added to report bug (GitHub issue #--- @---)
- [ ] Something not listed here

### Is this a breaking change? (select 1):
- [ ] 🚨 YES 🚨
- [X] No
- [ ] I'm not sure

### Is this ready to be merged into Dynamoose? (select 1):
- [X] Yes
- [ ] No

### Are all the tests currently passing on this PR? (select 1):
- [X] Yes
- [ ] No

### Other:
- [X] I have read through and followed the Contributing Guidelines
- [X] I have searched through the GitHub pull requests to ensure this PR has not already been submitted
- [ ] I have updated the Dynamoose documentation (if required) given the changes I made
- [ ] I have added/updated the Dynamoose test cases (if required) given the changes I made
- [ ] I have ensured the following commands are successful from the root of the project directory
  - [X] `npm test`
  - [X] `npm run lint`
- [X] I agree that all changes made in this pull request may be distributed and are made available in accordance with the [Dynamoose license](https://github.com/dynamoose/dynamoose/blob/master/LICENSE)
- [X] All of my commits and commit messages are detailed, explain what changes were made, and are easy to follow and understand
- [X] I have filled out all fields above

### CODE REVIEW

File: .eslintrc.js [hunk 1]
```diff
@@ -1,51 +1,41 @@
...
+    "@typescript-eslint/interface-name-prefix": "off",
+    "@typescript-eslint/explicit-function-return-type": "off",
+    "@typescript-eslint/no-explicit-any": "off",
+    "@typescript-eslint/no-unused-vars": "off",
+    "@typescript-eslint/camelcase": "off",
+    "@typescript-eslint/no-this-alias": "off",
+    "@typescript-eslint/no-var-requires": "off",
+    "@typescript-eslint/no-explicit-any": "off",
+    "object-curly-spacing": ["error", "never"],
+    "space-before-function-paren": ["error", "never"],
```
Comment (fishcharlie) [inline .eslintrc.js:27]: I think this should be always.

```suggestion
    "space-before-function-paren": ["error", "always"],
```

File: .eslintrc.js [hunk 2]
```diff
@@ -1,51 +1,41 @@
...
+  extends: [
+    "eslint:recommended",
+    "plugin:@typescript-eslint/eslint-recommended",
+    "plugin:@typescript-eslint/recommended",
+  ],
+  rules: {
+    "@typescript-eslint/interface-name-prefix": "off",
+    "@typescript-eslint/explicit-function-return-type": "off",
+    "@typescript-eslint/no-explicit-any": "off",
+    "@typescript-eslint/no-unused-vars": "off",
```
Comment (fishcharlie) [inline .eslintrc.js:21]: Why did we turn these off?

File: .eslintrc.js [hunk 3]
```diff
@@ -1,51 +1,41 @@
...
+    "@typescript-eslint/camelcase": "off",
+    "@typescript-eslint/no-this-alias": "off",
+    "@typescript-eslint/no-var-requires": "off",
+    "@typescript-eslint/no-explicit-any": "off",
+    "object-curly-spacing": ["error", "never"],
```
Comment (fishcharlie) [inline .eslintrc.js:26]: I actually do like the look of spaces before and after the import objects and deconstructing assignments. I don't think I like it in array's tho.

What happens if we set this to always? Kinda just wanna see how it looks.

File: .eslintrc.js [hunk 4]
```diff
@@ -1,51 +1,41 @@
...
+    "space-before-function-paren": ["error", "never"],
+    "space-before-blocks": ["error", "always"],
+    "no-multi-spaces": "error",
+    "space-in-parens": "error",
+    "@typescript-eslint/comma-spacing": "error",
```
Comment (fishcharlie) [inline .eslintrc.js:31]: Let's group all the TS rules together

File: package.json [hunk 5]
```diff
@@ -88,5 +90,16 @@
...
+      "pre-commit": "lint-staged",
+      "pre-push": "npm test"
+    }
+  },
+  "lint-staged": {
+    "*.{js,ts}": [
+      "eslint --fix"
+    ]
+  }
+}
```
Comment (fishcharlie) [inline package.json:105]:
```suggestion
}

```

File: publish/index.js [hunk 6]
```diff
@@ -16,6 +16,42 @@ const os = require("os");
 const npmFetch = require("npm-registry-fetch");
 let package = require("../package.json");
 
+
+
+async function checkCleanWorkingDir() {
```
Comment (fishcharlie) [inline publish/index.js:21]: Any reason why these got moved?

File: package.json [hunk 7]
```diff
@@ -14,6 +14,8 @@
     "chai": "^4.2.0",
     "chai-as-promised": "^7.1.1",
     "eslint": "^6.8.0",
+    "husky": "^4.2.5",
+    "lint-staged": "^10.2.2",
```
Comment (fishcharlie) [inline package.json:18]: Curious to hear your thoughts behind this. We used husky in the past, and personally I wasn't a big fan of it. But wanna hear your perspective on it.

File: package.json [hunk 8]
```diff
@@ -88,5 +90,16 @@
...
+  ],
+  "husky": {
+    "hooks": {
+      "pre-commit": "lint-staged",
+      "pre-push": "npm test"
+    }
+  },
+  "lint-staged": {
+    "*.{js,ts}": [
+      "eslint --fix"
```
Comment (fishcharlie) [inline package.json:102]: If we do decide to keep this why not just change this to `npm run lint:fix` to keep it all in one place?

File: package.json [hunk 9]
```diff
@@ -88,5 +90,16 @@
...
-}
+  ],
+  "husky": {
+    "hooks": {
+      "pre-commit": "lint-staged",
+      "pre-push": "npm test"
+    }
+  },
+  "lint-staged": {
+    "*.{js,ts}": [
```
Comment (fishcharlie) [inline package.json:101]: Is there a reason we have to define this? Why can't it be part of the `npm run lint` script? Why do we need to define it in two places?

File: .eslintrc.js (hunk 2 above)
Comment (pfulop) [reply .eslintrc.js:21]: `@typescript-eslint/no-explicit-any` is I believe from the previous eslint. Everything else
is based on [this](https://github.com/nestjs/nest/blob/master/.eslintrc.js) as suggested in https://github.com/dynamoose/dynamoose/issues/896. 

From my standpoint of view:

- interface-name-prefix: I really don't see reason in prefixing interfaces with I
- explicit-function-return-type: this is usually just an extra overhead for the developer from my experience Mainly considering typescript is really good with inferring return types now.
- no-explicit-any: I am personally not a fan of this but I think turning this on would slow down and make this PR big

File: .eslintrc.js (hunk 3 above)
Comment (pfulop) [reply .eslintrc.js:26]: I also like it in object and such, I based this rules on the comments from https://github.com/dynamoose/dynamoose/pull/890 but now I see there were two ways to interpret the comments and I probably chose wrong,

File: package.json (hunk 7 above)
Comment (pfulop) [reply package.json:18]: Husky can be a pain in the ass but I find it mostly helping in any kind of project. Most of the issues such as these (https://github.com/dynamoose/dynamoose/pull/890) won't happen with husky anymore (unless somebody skips commit prechecks). 

My thought is to always let tooling take any burden out of the developer's mind. I don't like to think about styling and usually forget about it when I am too focused on some issues. Then I submit PR and instead of moving forward with it I need somebody to correct me on my styling issues, which is just a slowdown and kinda annoying one. So if I can prevent it with anything I will. The argument would be to format files on save (which I also like to do) but for this project, my editor for some reason doesn't format according to the eslint rules.

TL;DR: I like to have consistent peace of mind for all developers that before they create PR their changes follow code styling of the project without much effort from their side.

File: package.json (hunk 9 above)
Comment (pfulop) [reply package.json:101]: I tried and it will also check package.json for instance.  Lint-staged supplies files to be checked and it will override `--ext .ts,.js `parameter for eslint.

**Comment by pfulop** [comment]
> How do we feel about turning on all for ESLint instead of recommended?

This currently produces a huge amount of eslint errors that are not auto-fixable. Also, in my personal opinion, it's a bit excessive.

File: .eslintrc.js (hunk 2 above)
Comment (fishcharlie) [reply .eslintrc.js:21]: You do have `@typescript-eslint/no-explicit-any` twice in here tho.

`interface-name-prefix` also didn't cause any problems before. Like we didn't have that disabled in the previous version and everything was working. What changed between last version and this version that required setting that to off?

`explicit-function-return-type`. I think that should be a separate pull request, and I'm not so sure I'd approve it. I don't see the harm in it as of right now, and as a newbie to TS, it does help get the syntax down and get in the habit of declaring return types.

`no-explicit-any`. Agreed. Separate PR would be good. However, it's duplicated twice in the rules. So we need to remove one of them.

File: .eslintrc.js [hunk 10]
```diff
@@ -1,51 +1,42 @@
...
+    "object-curly-spacing": ["error", "always"],
+    "space-before-function-paren": ["error", "always"],
+    "space-before-blocks": ["error", "always"],
+    "no-multi-spaces": "error",
+    "space-in-parens": "error",
+    indent: ["error", "tab"],
+    "linebreak-style": ["error", "unix"],
+    quotes: ["error", "double"],
+	semi: ["error", "always"],
+	"quote-props": ["error", "always"],
```
Comment (fishcharlie) [inline .eslintrc.js:36]: Formatting got messed up here. This is also something that ESLint should catch and throw an error about. So before fixing it. Please make sure that the linter currently throws an error here.

File: package.json (hunk 7 above)
Comment (fishcharlie) [reply package.json:18]: @pfulop And how about `lint-staged`? Never heard of that library before.

File: package.json [hunk 11]
```diff
@@ -88,5 +90,16 @@
...
   "files": [
     "dist"
-  ]
+  ],
+  "husky": {
+    "hooks": {
+      "pre-commit": "lint-staged",
+      "pre-push": "npm test"
+    }
+  },
```
Comment (fishcharlie) [inline package.json:99]: Personally. I prefer to have no configuration in the `package.json` file. It just adds complexity to understanding it. Let's either refactor this out into a different `.huskyrc.js` or `husky.config.js` file. I don't have a preference as to which we use. But also using a JS file will allow for comments, which might be helpful in the future.

File: package.json [hunk 12]
```diff
@@ -88,5 +90,16 @@
   },
   "files": [
     "dist"
-  ]
+  ],
+  "husky": {
+    "hooks": {
+      "pre-commit": "lint-staged",
```
Comment (fishcharlie) [inline package.json:96]: So if this gets run before commit. It runs `npm run lint:fix`, however, does it actually stage those files? Or would it run the lint, fix the files, commit, then still have the lint changes in the working copy?

**Comment by fishcharlie** [comment]
Also, we got merge conflicts.

File: package.json (hunk 12 above)
Comment (pfulop) [reply package.json]: `lint-staged` is a library that runs the specified command (in this case linting) only on staged files, therefore, your PR will have only stuff that you changed (because if linter would run globally and somebody would push without checks, you might style fix his file and it would end up in your PR).

`lint-staged` will also stage files that got fixed by linter that was run by husky. Maybe that's why you didn't like husky because it was used without lint-staged.

File: .eslintrc.js (hunk 10 above)
Comment (pfulop) [reply .eslintrc.js]: Added .eslintrc.js to be explicitly included since by default it's ignored by eslint.

Comment (pfulop) [reply .eslintrc.js]: `interface-name-prefix` yeah I guess it's not needed in this project since interface names are following typescript recommendations.

`explicit-function-return-type` by default it's just a warning. I would let it go through this PR turned off and then maybe create another one for turning it into error if you really think it will help you. Otherwise, in my opinion, it's too much pain for a (mostly none) little gain.

File: .eslintignore [hunk 13]
```diff
@@ -1,3 +1,4 @@
 docs
 dist
 publish/node_modules
+!.eslintrc.js
```
Comment (fishcharlie) [inline .eslintignore:4]: New line at end of files.

```suggestion
!.eslintrc.js

```

File: .eslintrc.js [hunk 14]
```diff
@@ -1,51 +1,40 @@
...
-			}
-		}
+	"plugins": ["@typescript-eslint/eslint-plugin"],
+	"extends": [
+		"eslint:recommended",
+		"plugin:@typescript-eslint/eslint-recommended",
+		"plugin:@typescript-eslint/recommended",
 	],
+	"rules": {
+		"@typescript-eslint/explicit-function-return-type": "off",
```
Comment (fishcharlie) [inline .eslintrc.js:18]: Gonna make the call to remove this. We can discuss adding it back later.

```suggestion
```

File: .huskyrc.js [hunk 15]
```diff
@@ -0,0 +1,6 @@
+module.exports = {
+  hooks: {
```
Comment (fishcharlie) [inline .huskyrc.js:2]: Hmm. This should have failed the linter since it doesn't have quotes around it. Any ideas why it didn't?

```suggestion
  "hooks": {
```

File: package.json (hunk 5 above)
Comment (fishcharlie) [reply package.json:94]: This still needs to be done.

**Comment by pfulop** [comment]
I had to go back to eslint command inside lint-staged instead of npm run lint:fix since that one doesn't accept files as an input.

**Comment by fishcharlie** [comment]
> I had to go back to eslint command inside lint-staged instead of npm run lint:fix since that one doesn't accept files as an input.

@pfulop Ahh. We might need to add `--` in front of it or something. Do you have `Allow edits from maintainers` enabled? If so, mind if I push directly to your branch so I can finish this up?

**Comment by pfulop** [comment]
I've sent you an invite (not sure where Allow edits from maintainers setting is). Feel free to push what you need into the branch

**Comment by fishcharlie** [comment]
@pfulop Wanna review #913 & #914? Basically took what you did in this pull request, spilt it out into two different pull requests, and made a few more minor changes. But **really** heavily based on this pull request.

Thank you for your hard work on this. It would have taken me much longer to do those pull requests if I wasn't able to copy and paste from this one.
